src/context.rs

```diff
@@ -7,6 +7,19 @@ use Secp256k1;
 #[cfg(feature = "std")]
 pub use self::std_only::*;
 
+/// A type that represents the type with the biggest alignment we can get in rust.
+/// Trying to match what `malloc` does in C, this should be aligned enough to contain pointers too.
+/// This type can have different size/alignment depending on the architecture.
+#[cfg(any(target_pointer_width = "32", target_pointer_width = "16", target_pointer_width = "8"))]
+pub type AlignType = u64;
```

Comment:
Uh, based on what documentation are you assuming that this is a type with "the biggest alignment we can get"? I don't think such a type can even exist, since we can do things like
```rust
#[repr(align(256))]
struct A(u8);
```
This type is [256-aligned](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=ced5aa3565213e2cf9d5f0196c760379).

I am not sure what you are trying to do here, but it seems extremely fishy.

Reply:
I'll start with the purpose and then with "the biggest".
The purpose is creating a pointer that is valid for a C function contract that requires a pointer from `malloc`:

>  The pointer returned if the allocation succeeds is suitably aligned so that it may be assigned to a pointer to any type of object with a fundamental alignment requirement and then used to access such an object or an array of such objects in the space allocated (until the space is explicitly deallocated).

Now `malloc` doesn't define what is this. and currently pointers in 64bit architecture are `u64` so we know that we need *at least* this much alignment because the C library does write pointers into that memory.
The only thing currently that can align more than that is `__uint128_t` which is 16 bytes alignment.

Now to why this value was chose.
we support rust 1.22.0 here (last point: https://github.com/rust-bitcoin/rust-secp256k1#contributing) so we can't use `#[repr(align(N))]`, and even if we could i'm not sure what would be the right value to put there.
but currently as we don't have support for that, and we know that we need at most 8 bytes alignment we use `u64` for alignment. 
I agree that's not a good solution, but because rust will never have `max_align_t` and we don't have access to C's max alignment (as it's not defined AFAIK), so currently the only thing we can do is maintain an alignment as big as the biggest primitive's alignment that we can check.

Reply:
Okay, so what you really want is C's `alignof(max_align_t)`? The Rust libstd has [a table for this](https://github.com/rust-lang/rust/blob/5670d048c0f88af9976b5505c7853b23dd06770d/src/libstd/sys_common/alloc.rs#L7-L23) that I [copied into Miri](https://github.com/rust-lang/miri/blob/44ca0f6a35e62ac9d317f3b0162eba56b1118137/src/shims/foreign_items.rs#L18-L22) at some point. I don't think it agrees with the value you have been using here, though. In particular, on 64bit x86, `max_align_t` is 16-byte aligned.

However, how far does this "requires a pointer from malloc" go? Does the pointer have to be freeable with `free`? If yes, you *have to* call `malloc`; the pointer you get from Rust's alloactor might or might not come from `malloc`. Even if no, it might be easiest to just call `malloc`?

@gnzlbg does `libc` expose `max_align_t`, or is there some other way to get that information though `libc`?

I am still confused why you need this as a *type* though, as opposed to just having to know the alignment and then using that (together with the size, which I assume is supplied separately) when allocating.

Reply:
> @gnzlbg does libc expose max_align_t, or is there some other way to get that information though libc?

Not yet but this seems a reasonable thing to implement.

Reply:
> I am still confused why you need this as a type though, as opposed to just having to know the alignment and then using that (together with the size, which I assume is supplied separately) when allocating.

I suppose that people wanted to be able to use C's `alignof(<type>)` to compute the maximing alignment, so instead of creating a `MAX_ALIGN` constant, they decided to over engineer it and make it a type. 

If we expose this from libc, in Rust, you can just do `const MAX_ALIGN: usize = align_of::<libc::max_align_t>();` and use that. In `jemalloc-sys` we use a hardcoded `MAX_ALIGN` constant for that.

Reply:
@RalfJung the exact definition from upstream is:

> The caller must provide a pointer to a rewritable contiguous block of memory of size at least secp256k1_context_preallocated_size(flags) bytes, suitably aligned to hold an object of any type.
So being able to pass it to `free()` isn't a requirement, being aligned as max is.
https://github.com/bitcoin-core/secp256k1/blob/master/include/secp256k1_preallocated.h#L41

Reply:
> is it possible to achieve alignment 16 without repr(align(N))?

`core::arch::...::__m128` have alignment of 16.

Reply:
> the exact definition from upstream is

I see. "object of any type" is certainly not true even in C with compiler extensions, as C also has a way to increase alignment similar to `#[repr(align)]` in Rust. So *probably* upstream means `max_align_t` but it might be a good idea to ask them to clarify their docs.

So, why do you need a *type* that has that alignment, as opposed to just using [`alloc::alloc`](https://doc.rust-lang.org/nightly/std/alloc/fn.alloc.html)? Is Rust 1.28 (more than a year old at this point!) too "new"?

Reply:
> "object of any type" is certainly not true even in C with compiler extensions,

That's because that's incorrect. The requirement is that the allocation must be suited for any type with **fundamental** alignment, where "fundamental" means that you did not use anything like `repr(align())` anywhere on it, among other things. 

If you use `repr(align())` on it in such a way that the alignment of the type is larger than the largest alignment of such a "fundamental" type, then you **can't** use `malloc` to allocate its memory, and need to use an API that accepts an alignment instead.

Reply:
You're right when it comes to modern C.
But upstream uses C89, so there's no `max_align_t` and no `_Alignas`, and AFAIK there's no sense of "fundamental type".
But it still may be confusing so i'll ask upstream what people think. Thanks.